packages/concerto-core/lib/decoratorutil.js

```diff
@@ -0,0 +1,371 @@
+                dictVoc[decl.declaration][extensionKey] = dcs.arguments[0].value;
+            }
+        }
+        return dictVoc;
+    }
+    /**
+    * Adds a key-value pair to a dictionary (object) if the key exists,
+    * or creates a new key with the provided value.
+    *
+    * @param {Object} dictionary - The dictionary (object) to which to add the key-value pair.
+    * @param {string} key - The key to add or update.
+    * @param {any} value - The value to add or update.
+    * @param {Object} options - options containing target
+    * @param {string} options.declaration - Target declaration
+    * @param {string} options.property - Target property
+    * @param {string} options.mapElement - Target map element
+    * @returns {Object} - constructed DCS Dict
+    * @private
+    */
+    static constructDCSDictionary(dictionary, key, value, options) {
+        const val = {
+            declaration:options?.declaration || '',
+            property:options?.property || '',
+            mapElement:options?.mapElement || '',
+            dcs: JSON.stringify(value),
+        };
+        if (dictionary[key] && Array.isArray( dictionary[key])) {
+            dictionary[key].push(val);
+        } else {
+            dictionary[key] = [val];
+        }
+        return dictionary;
+    }
+    /**
+    * parses the extracted decorators and generates arrays of decorator command set and vocabularies
+    *
+    * @param {Object} extractionDictionary - extracted decorators and vocabularies
+    * @param {String} DCS_VERSION - dcs_version
+    * @param {String} locale - locale for vocabularies
+    * @returns {Object} - constructed DCS Dict and processed models ast
+    * @private
+    */
+    static parseDecosAndVocabs(extractionDictionary, DCS_VERSION, locale){
+        const decoratorData = [];
+        const vocabData = [];
+        Object.keys(extractionDictionary).forEach(namespace => {
+            const {name, version} = ModelUtil.parseNamespace(namespace);
+            const nameOfDcs = name;
+            const versionOfDcs = version;
+            const jsonData = extractionDictionary[namespace];
+            const patternToDetermineVocab = /^Term_/i;
+            let dcsObjects = [];
+            let vocabObject = {};
+            jsonData.forEach(obj =>{
+                const decos = JSON.parse(obj.dcs);
+                const target = {
+                    '$class': `org.accordproject.decoratorcommands@${DCS_VERSION}.CommandTarget`,
+                    'namespace':namespace
+                };
+                if (obj.declaration && obj.declaration!==''){
+                    target.declaration = obj.declaration;
+                }
+                if (obj.property && obj.property!==''){
+                    target.property = obj.property;
+                }
+                if (obj.mapElement && obj.mapElement!==''){
+                    target.mapElement = obj.mapElement;
+                }
```

Comment:
Much cleaner now, but since `target` is only needed within `parseNonVocabularyDecorators`, and never used else where down the line, in `parseNonVocabularyDecorators `, you can build `target` object within `parseNonVocabularyDecorators` and you can pass `DCS_VERSION` and `namespace` as options to `parseNonVocabularyDecorators`. That would make this loop a lot cleaner and contain the logic related to `parseNonVocabularyDecorators` in one place.

Reply:
Since the `parseNonVocabularyDecorators` is called per declaration level, a new target will be getting created for each declaration. however, currently target declaration is outside the declaration loop. I think it is more performant currently.
Do you think I should still move it to `parseNonVocabularyDecorators` ?

Reply:
Ah, in that case, it doesn't make sense to move inside the function, but maybe move it inside a separate function called, `buildDecoratorTarget` or something? Here it just crowds the rest of the function.

Reply:
updated.